rust/automerge/src/iter/doc.rs

```diff
@@ -81,7 +62,132 @@ pub(crate) struct DocIterInternal<'a> {
     obj: ObjId,
 }
 
+#[derive(Debug, Clone)]
+pub(crate) struct DiffIter<'a> {
```

Comment:
It bothers me a little bit that `DiffIter` and `DocIterInternal` have so much similar structure and logic which we're duplicating here. It's not obvious to me that it would be worth abstracting over in any way though, it might just be coincidental that the logic is shared for now and maybe we expect it to drift?